lib/ioc_container.dart

```diff
@@ -254,37 +260,55 @@ extension IocContainerExtensions on IocContainer {
       IocContainer(
         serviceDefinitionsByType,
         useExistingSingletons ? Map<Type, Object>.from(singletons) : {},
+        {},
         isScoped: true,
       );
 
   ///⌛ Gets a service that requires async initialization. Add these services
   ///with [IocContainerBuilder.addAsync] or
-  ///[IocContainerBuilder.addSingletonAsync] You can only use this on factories
+  ///[IocContainerBuilder.addSingletonAsync]. You can only use this on factories
   ///that return a Future<>.
-  ///Warning: if the definition is singleton/scoped and the Future fails, the factory will never return a
-  ///valid value, so use [getAsyncSafe] to ensure the container doesn't store
-  ///failed singletons
-  Future<T> getAsync<T>() async => get<Future<T>>();
-
-  ///See [getAsync].
-  ///Makes an async call by creating a temporary scoped container,
-  ///attempting to make the async initialization and merging the result with the
-  ///current container if there is success.
-  ///
-  ///⚠️ Warning: allows reentrancy and does not do error handling.
-  ///If you call this more than once in parallel it will create multiple
-  ///Futures - i.e. make multiple async calls. You need to guard against this
-  ///and perform retries on failure. Be aware that this may happen even if
-  ///you only call this method in a single location in your app.
-  ///You may need a an async lock.
-  Future<T> getAsyncSafe<T>() async {
-    final scope = scoped();
-
-    final service = await scope.getAsync<T>();
-
-    merge(scope);
+  Future<T> getAsync<T>() async {
+    final serviceDefinition = serviceDefinitionsByType[Future<T>];
 
-    return service;
+    if (serviceDefinition == null) {
+      throw ServiceNotFoundException<T>(
+        'Service $T not found',
+      );
+    }
+
+    if (serviceDefinition.isSingleton || isScoped) {
+      final singletonValue = singletons[Future<T>];
+
+      if (singletonValue != null) {
+        //Return completed successful future
+        return singletonValue as Future<T>;
+      }
+
+      if (!locks.containsKey(T)) {
+        //Add a lock
+        locks[T] =
+            AsyncLock<T>(() => serviceDefinition.factory(this) as Future<T>);
+      }
+
+      final lock = locks[T]! as AsyncLock<T>;
+
+      try {
+        //Await the locked call
+        final future = lock.execute();
+        await future;
+
+        //Store successful future
+        singletons[Future<T>] = future;
+
+        return future;
```

Comment:
Clever. I learnt something new with `await`ing on a `Future` that has already been `await`ed